internal/catalogmetadata/cache/cache_test.go

```diff
@@ -58,291 +52,113 @@ const (
-			catalog: &catalogd.ClusterCatalog{
-				ObjectMeta: metav1.ObjectMeta{
-					Name: "test-catalog",
-				},
-				Status: catalogd.ClusterCatalogStatus{
-					ResolvedSource: &catalogd.ResolvedCatalogSource{
-						Image: nil,
-					},
-				},
-			},
-			contents: defaultFS,
-			tripper:  &mockTripper{serverError: true},
-			wantErr:  true,
-		},
-	} {
-		t.Run(tt.name, func(t *testing.T) {
-			ctx := context.Background()
-			cacheDir := t.TempDir()
-			tt.tripper.content = make(fstest.MapFS)
-			maps.Copy(tt.tripper.content, tt.contents)
-			httpClient := &http.Client{
-				Transport: tt.tripper,
-			}
-			c := cache.NewFilesystemCache(cacheDir, func() (*http.Client, error) {
-				return httpClient, nil
-			})
-
-			actualFS, err := c.FetchCatalogContents(ctx, tt.catalog)
-			if !tt.wantErr {
-				assert.NoError(t, err)
-				assert.NoError(t, equalFilesystems(tt.contents, actualFS))
-			} else {
-				assert.Error(t, err)
-			}
+}
 
-			if tt.testCaching {
-				if !tt.shouldHitCache {
-					tt.catalog.Status.ResolvedSource.Image.ResolvedRef = "fake/catalog@sha256:shafake"
-				}
-				tt.tripper.content["foobar/olm.package/foobar.json"] = &fstest.MapFile{Data: []byte(`{"schema": "olm.package", "name": "foobar"}`)}
-				actualFS, err := c.FetchCatalogContents(ctx, tt.catalog)
-				assert.NoError(t, err)
-				if !tt.shouldHitCache {
-					assert.NoError(t, equalFilesystems(tt.tripper.content, actualFS))
-					assert.ErrorContains(t, equalFilesystems(tt.contents, actualFS), "foobar/olm.package/foobar.json")
-				} else {
-					assert.NoError(t, equalFilesystems(tt.contents, actualFS))
-				}
-			}
-		})
-	}
+func TestFilesystemCachePutAndGet(t *testing.T) {
+	const (
+		catalogName  = "test-catalog"
+		resolvedRef1 = "fake/catalog@sha256:fakesha1"
+		resolvedRef2 = "fake/catalog@sha256:fakesha2"
+	)
+
+	cacheDir := t.TempDir()
+	c := cache.NewFilesystemCache(cacheDir)
+
+	catalogCachePath := filepath.Join(cacheDir, catalogName)
+	require.NoDirExists(t, catalogCachePath)
+
+	t.Log("Get empty v1 cache")
+	actualFSGet, err := c.Get(catalogName, resolvedRef1)
+	assert.NoError(t, err)
+	assert.Nil(t, actualFSGet)
+
+	t.Log("Put v1 content into cache")
+	actualFSPut, err := c.Put(catalogName, resolvedRef1, defaultContent(), nil)
+	assert.NoError(t, err)
+	assert.NoError(t, equalFilesystems(defaultFS(), actualFSPut))
+
+	t.Log("Get v1 content from cache")
+	actualFSGet, err = c.Get(catalogName, resolvedRef1)
+	assert.NoError(t, err)
+	assert.NoError(t, equalFilesystems(defaultFS(), actualFSPut))
+	assert.NoError(t, equalFilesystems(actualFSPut, actualFSGet))
+
+	t.Log("Put v1 error into cache")
+	actualFSPut, err = c.Put(catalogName, resolvedRef1, nil, errors.New("fake put error"))
+	// Errors do not override previously successfully populated cache
+	assert.NoError(t, err)
+	assert.NoError(t, equalFilesystems(defaultFS(), actualFSPut))
+	assert.NoError(t, equalFilesystems(actualFSPut, actualFSGet))
+
+	t.Log("Put v2 error into cache")
```

Comment:
Should we add a test after populating something for a newly resolved ref that doing a GET for the old resolved ref returns a nil cache and nil error (IIUC that is the intended behavior)?

Reply:
Actually you are right - "Get v1 from cache" will return `nil` because we map by catalog name and override on `Put`. This is a bug: I want `Remove` to work even if we do a `Put` of another version before `Remove`.

I think #1207 (and `main`) has the same issue...